### DISCUSSION TITLE
Figure out how to strip config of persisted credentials from checkout action

### DISCUSSION THREAD

**Discussion by JamesIves** [JamesIves/github-pages-deploy-action#506] [Ideas]
Right now you must run `checkout@v2` with `persist-credentials: false` in order for the deployment to follow through correctly if you're using an access token variable. This is confusing, and has [caused some discussion](https://github.com/JamesIves/github-pages-deploy-action/issues/389#issuecomment-730042615) in the past.

It would be great if the action could tear down any pre-set variables from previous action runs so it starts on a fresh foot. The action already manages its own credentials so it should safely toss out any ones provided by others.

This is what `persist-credentials` actually does: https://github.com/actions/checkout/blob/main/src/git-auth-helper.ts#L137-L168

I'm open to discussions/ideas on how to prevent this and I've created an issue for this here: https://github.com/JamesIves/github-pages-deploy-action/issues/507

**Comment by Pike** [comment]
Is this just for access tokens? SSH and GITHUB_TOKEN are fine?

I'm asking because I wanted to propose to drop the `ACCESS_TOKEN` feature. Happy to open a new post here to gather feedback. My concern is that it's very easy to create these tokens, and much harder to do so correctly. And I would think that one could generally use either a deploy key or just the GITHUB_TOKEN?

**Reply by JamesIves** [reply]
This affects the access token specifically, as it will result in the GitHub token taking precedence if persist-credentials is not false. The user has to specifically provide some form of token for the deploy action to follow through, and this creates a degree of inconsistency as the token they provide will be overwritten by something else entirely as it gets baked into the configuration.

There's been some shakiness in the past with the GitHub token secret not providing adequate access in private repositories (it flat out just didn't work at all for a long time and didn't trigger a page build). Using an access token is also required if you are deploying cross repository/organization. I think all we can do is steer the user in the right direction. As there's a genuine use case for needing an access token we shouldn't drop support for it. It's definitely a nuclear option though I agree.

**Reply by JamesIves** [reply]
Some additional thoughts on this.

I think it's probably a good idea to remove the `GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}` variable all together in version 4. I think we should only be switching the deployment method if `ACCESS_TOKEN` or `SSH` is specified, and default to GITHUB_TOKEN.  This is the approach a lot of the GitHub provided actions are taking now too as it's available to default in the `action.yml` file.

**Reply by Pike** [reply]
I like it.

Would that imply that if the checkout action uses SSH, the deploy action would do so, too?

**Reply by JamesIves** [reply]
No because the deploy key is not provided by default in all workflows. This will only be applicable to the Github token as it's always available. This just simplifies the boilerplate for 90% of workflows.

**Reply by JamesIves** [reply]
PR: https://github.com/JamesIves/github-pages-deploy-action/pull/530

**Reply by emmahsax** [reply]
Personally, I don't like using the `access_token`. Because personal access tokens are valid for all of a GitHub user's repositories, it is allowed to do many things, and that seems slightly more dangerous to me than using `secrets.GITHUB_TOKEN` or SSH keys. Therefore, I'd be cool with dropping support for access tokens, but I also understand how some people might want to use that because then they can use one token for many projects.

In a different  note, I'd be fine with it defaulting to `secrets.GITHUB_TOKEN` though.
